src/robosar_gui.py

```diff
@@ -1,33 +1,112 @@
+from PyQt5 import QtWidgets, QtGui
+from PyQt5.QtCore import QTimer, QSize
+from std_msgs.msg import String, Bool, Int32
+from sensor_msgs.msg import Image
 from robosar_messages.srv import *
 from robosar_messages.msg import *
+from gui_designer import Ui_Dialog
+
+ALIVE = ["ROBOT_STATUS_ACTIVE", "ROBOT_STATUS_INACTIVE"]
+DEAD = ["ROBOT_STATUS_COMM_FAIL",
+        "ROBOT_STATUS_UNREACHABLE", "ROBOT_STATUS_NO_HEARTBEAT"]
+
+
+class AgentGroup():
+    def __init__(self) -> None:
+        self.group_box = None
+        self.grid_layout = None
+        self.status_label = None
+        self.battery_label = None
+        self.feedback_label = None
+        self.ip_label = None
 
 
 class Ui(QtWidgets.QDialog):
     def __init__(self):
         super(Ui, self).__init__()
+        self.ui = Ui_Dialog()
+        self.ui.setupUi(self)
+
+        # key: agent name, value: Bool
+        self.agent_active_status = {}
+        # key: agent name, value: AgentGroup
+        self.agent_status_dict = {}
+
+        # init ROS stuff
         rospy.init_node('robosar_gui')
-        self.dummy_pub = rospy.Publisher('dummy_topic', String, queue_size=1)
+        rospy.Subscriber("/task_allocation_image", Image,
+                         self.display_task_allocation)
         rospy.Subscriber("/robosar_agent_bringup_node/status",
                          Bool, self.get_active_agents)
-        uic.loadUi(
-            os.path.join(os.path.dirname(os.path.abspath(__file__)), 'robosar_gui.ui'), self)
-        self.agent_active_status = {}
-        self.button = self.findChild(
-            QtWidgets.QPushButton, 'pushButton')  # Find the button
-        self.agent_status_text = self.findChild(QtWidgets.QLabel, 'label')
-        self.button.clicked.connect(self.publish_button_pressed)
         self.get_active_agents()
+        rospy.Subscriber("/robosar_agent_bringup_node/all_agent_status",
+                         agents_status, self.display_agents_status)
+        rospy.Subscriber("/tasks_completed", Int32, self.display_task_count)
+
+        self.start_time = 0.0
+        self.elasped_time = 0
+        self.start = False
+        self.ui.start_mission_button.clicked.connect(self.send_mission)
+        self.ui.start_timer_button.clicked.connect(self.start_timer)
+        self.ui.stop_timer_button.clicked.connect(self.pause_timer)
+        self.ui.restart_timer_button.clicked.connect(self.restart_timer)
+        self.agent_scroll_area = self.ui.scrollArea
+
+        self.mission_timer = QTimer(self)
+        self.mission_timer.timeout.connect(self.show_time)
 
         self.show()
 
+    def display_task_count(self, msg):
+        self.ui.tasks_completed_label.setText(str(msg.data))
+
+    def display_task_allocation(self, msg):
+        if msg:
+            br = CvBridge()
+            try:
+                data = br.imgmsg_to_cv2(msg, "rgb8")
+                height, width, _ = data.shape
+                bytesPerLine = 3 * width
```

Comment:
Can you add any comment regarding how you got the value for bytesPerLine?

Reply:
google